File: Scala/src/main/scala/AlgorithmicHeights/SCC.scala
package AlgorithmicHeights

object SCC {
  import scala.annotation.tailrec
  import scala.collection.mutable.ListBuffer

  private type Component = List[Int]
  final case class Edge(a: Int, b: Int)
  final case class DFSResult(components: List[Component],
                             topologicalOrdering: List[Int],
                             previsitIds: List[Int],
                             postvisitIds: List[Int])

  class DirectedGraph(val nrNodes: Int, edgeList: List[Edge], nodeOrder: Option[List[Int]] = None) {
    import DirectedGraph.buildAdjacencyList

    val orderedNodes: List[Int] = nodeOrder match {
      case None => (1 to nrNodes).toList
      case Some(nodes) => nodes
    }

    private val adjacencyList: Map[Int, List[Int]] = buildAdjacencyList(edgeList)

    def neighbours(node: Int): List[Int] = adjacencyList.getOrElse(node, Nil)
  }

  object DirectedGraph {
    private def buildAdjacencyList(edgeList: List[Edge]): Map[Int, List[Int]] =
      edgeList.groupMap{ case Edge(a, _) => a }{ case Edge(_, b) => b }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def readParameters(line: String): (Int, Int) = convertToIntList(line) match {
    case List(n, e) => (n, e)
    case _ => throw new Exception("Malformed input.")
  }

  private def readEdges(reader: Iterator[String], nrEdges: Int): List[Edge] =
    reader
      .take(nrEdges)
      .map(convertToIntList)
      .collect{ case List(a, b) => Edge(a, b) }
      .toList

  private def depthFirstSearch(graph: DirectedGraph): DFSResult = {
    val visitStarted: Array[Int] = Array.fill(graph.nrNodes)(0)
    val visitEnded: Array[Int] = Array.fill(graph.nrNodes)(0)
    val topologicalSorting: ListBuffer[Int] = ListBuffer()
    val previsitId: Iterator[Int] = Iterator.from(1)
    val postvisitId: Iterator[Int] = Iterator.from(1)

    def isNodeVisited(node: Int): Boolean = visitStarted(node - 1) > 0

    def findUnvisitedNeighbour(node: Int): Option[Int] = graph.neighbours(node).find(!isNodeVisited(_))

    def explore(startingNode: Int): Component = {
      @tailrec
      def traverseComponent(previsitStack: List[Int], component: Component): Component = previsitStack match {
        case Nil => component
        case node :: restOfStack =>
          findUnvisitedNeighbour(node) match {
            case Some(neighbour) =>
              visitStarted(neighbour - 1) = previsitId.next()
              traverseComponent(neighbour :: previsitStack, neighbour :: component)
            case None =>
              visitEnded(node - 1) = postvisitId.next()
              topologicalSorting += node
              traverseComponent(restOfStack, component)
          }
      }

      visitStarted(startingNode - 1) = previsitId.next()
      traverseComponent(List(startingNode), List(startingNode))
    }

    @tailrec
    def findComponents(nodes: List[Int], components: List[Component]): List[Component] = nodes match {
      case Nil => components
      case node :: remainingNodes =>
        if (isNodeVisited(node)) findComponents(remainingNodes, components)
        else {
          val currentComponent: Component = explore(node)
          findComponents(remainingNodes, currentComponent :: components)
        }
    }

    val components: List[Component] = findComponents(graph.orderedNodes, Nil)
    DFSResult(components, topologicalSorting.reverseIterator.toList, visitStarted.toList, visitEnded.toList)
  }

  private def createGraphWithEdgesReversed(nrNodes: Int, edges: List[Edge], postvisitIds: List[Int]): DirectedGraph = {
    val reversedEdges: List[Edge] = edges.map{ case Edge(from, to) => Edge(to, from) }
    val nodeOrder: List[Int] =
      (1 to nrNodes)
        .zip(postvisitIds)
        .sortBy{ case (_, postvisitId) => postvisitId }(Ordering[Int].reverse)
        .map{ case (id, _) => id }
        .toList
    new DirectedGraph(nrNodes, reversedEdges, Some(nodeOrder))
  }

  def calcStronglyConnectedComponents(nrNodes: Int, edges: List[Edge]): List[Component] = {
    val graph = new DirectedGraph(nrNodes, edges)
    val DFSResult(_, _, _, postvisitNumbers) = depthFirstSearch(graph)
    val reversedGraph: DirectedGraph = createGraphWithEdgesReversed(nrNodes, edges, postvisitNumbers)
    val DFSResult(components, _, _, _) = depthFirstSearch(reversedGraph)
    components
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (nrNodes, nrEdges): (Int, Int) = readParameters(reader.next())
    val edgeList: List[Edge] = readEdges(reader, nrEdges)
    val components: List[Component] = calcStronglyConnectedComponents(nrNodes, edgeList)
    println(components.length)
  }
}

File: OCaml/src/algorithmic_heights/scc.ml
module IntMap = Map.Make(Int)

type edge = { a : int; b : int }
type component = int list
type dfsresult = { components : component list; previsit_ids : int list; postvisit_ids : int list }

module Graph : sig
    type t
    val dfs : t -> int list option -> dfsresult
    val create : int -> edge list -> t
end = struct
    type t = { nr_nodes : int; adjacency_list : int list IntMap.t }

    let get_neighbors adjacency_list node =
        match IntMap.find_opt node adjacency_list with
            | None -> []
            | Some neighbors -> neighbors

    let create nr_nodes edges =
        let add_edge adjacency_list { a; b } =
            let update_function = function
                | None -> Some [b]
                | Some ns -> Some (b :: ns) in
            IntMap.update a update_function adjacency_list in
        let adjacency_list = List.fold_left add_edge IntMap.empty edges in
        { nr_nodes; adjacency_list }

    let dfs { nr_nodes; adjacency_list } nodes =
        let visit_started = Array.make nr_nodes 0 in
        let visit_ended = Array.make nr_nodes 0 in
        let previsit_id = ref 1 in
        let postvisit_id = ref 1 in
        let is_node_visited node = visit_started.(node - 1) > 0 in
        let find_unvisited_neighbor node =
            node |> get_neighbors adjacency_list |> List.find_opt (fun item -> not (is_node_visited item)) in

        let explore start_node =
            let rec traverse_component component = function
                | [] -> component
                | (node :: rest) as previsit_stack ->
                    match find_unvisited_neighbor node with
                        | Some neighbor ->
                            visit_started.(neighbor - 1) <- !previsit_id;
                            previsit_id := !previsit_id + 1;
                            traverse_component (neighbor :: component) (neighbor :: previsit_stack)
                        | None ->
                            visit_ended.(node - 1) <- !postvisit_id;
                            postvisit_id := !postvisit_id + 1;
                            traverse_component component rest in
            visit_started.(start_node - 1) <- !previsit_id;
            previsit_id := !previsit_id + 1;
            traverse_component [start_node] [start_node] in

        let rec find_components components = function
            | [] -> components
            | node :: remaining_nodes ->
                if is_node_visited node then find_components components remaining_nodes
                else
                    let current_component = explore node in
                    find_components (current_component :: components) remaining_nodes in

        let ordered_nodes = match nodes with
            | None -> List.init nr_nodes (fun k -> k + 1)
            | Some ns -> ns in
        let components = find_components [] ordered_nodes in
        { components = components
        ; previsit_ids = Array.to_list visit_started
        ; postvisit_ids = Array.to_list visit_ended }
end


let read_edges (nr_edges : int) : edge list =
    let parse_line line =
        Scanf.sscanf line "%d %d" (fun a b -> { a; b }) in
    let range = Seq.init nr_edges (fun k -> k) in
    List.of_seq @@ Seq.map (fun _ -> parse_line (read_line ())) range


let get_node_visit_order_in_reversed_graph (postvisit_ids : int list) : int list =
    let compare_pairs (_, id1) (_, id2) = compare id2 id1 in
    List.(postvisit_ids |> mapi (fun node id -> (node + 1, id)) |> sort compare_pairs |> map (fun (node, _) -> node))


let calc_strongly_connected_components (nr_nodes : int) (edges : edge list) : component list =
    let graph = Graph.create nr_nodes edges in
    let { postvisit_ids; _ } = Graph.dfs graph None in
    let reversed_edges = List.map (fun {a; b} -> {a = b; b = a}) edges in
    let reversed_graph = Graph.create nr_nodes reversed_edges in
    let node_order = get_node_visit_order_in_reversed_graph postvisit_ids in
    let { components; _ } = Graph.dfs reversed_graph (Some node_order) in
    components


let () =
    let (nr_nodes, nr_edges) = Scanf.sscanf (read_line ()) "%d %d" (fun n e -> (n, e)) in
    let edge_list = read_edges nr_edges in
    let components = calc_strongly_connected_components nr_nodes edge_list in
    print_int (List.length components); print_newline ()
